Lib/test/test_builtin.py

```diff
@@ -832,6 +832,31 @@ class customdict(dict):  # this one should not do anything fancy
         self.assertRaisesRegex(NameError, "name 'superglobal' is not defined",
                                exec, code, {'__builtins__': customdict()})
 
+    def test_exec_builtins_mapping(self):
+        code = compile("superglobal", "test", "exec")
+        # works correctly
+        exec(code, {'__builtins__': types.MappingProxyType({'superglobal': 1})})
+        # custom builtins mapping is missing key
+        ns = {'__builtins__': types.MappingProxyType({})}
+        self.assertRaisesRegex(NameError, "name 'superglobal' is not defined",
+                               exec, code, ns)
+
+    def test_exec_builtins_mapping_import(self):
+        code = compile("import foo.bar", "test", "exec")
+        ns = {'__builtins__': types.MappingProxyType({})}
+        self.assertRaisesRegex(ImportError, "__import__ not found", exec, code, ns)
+        ns = {'__builtins__': types.MappingProxyType({'__import__': lambda *args: args})}
+        exec(code, ns)
+        self.assertEqual(ns['foo'], ('foo.bar', ns, ns, None, 0))
+
+    def test_eval_builtins_mapping_reduce(self):
+        code = compile("iter([1, 2]).__reduce__()", "test", "eval")
```

Comment:
The test name is misleading. The test is more about getting a built-in function from custom builtins module (namespace), no?

I suggest testing that the code works as example. Such as:

```py
code = compile("x = str(123)", "test", "exec")
ns = {}
globals = {'__builtins__': types.MappingProxyType({'str': str})}
exec(code, globals, ns)
self.assertEqual(ns['x'], "123")
```

Reply:
It's testing `__reduce__` on a list iterator specifically because the implementation of `__reduce__` accesses `PyEval_GetBuiltins`. Your test wouldn't do that.

Reply:
The code already calls directly `iter()` explicitly. So it's not obvious if the expected error comes from this direct `iter()` call, or from the ``__reduce__()`` method call.

At least, it would be worth it to add a comment to explain these implementation details which are non obvious to me.

Reply:
That's a good point, might be better to pass `iter([1, 2])` as one of the entries in the builtins.

Reply:
Done.